[MIPS] Add Scheduling model for MIPS i6400 and i6500 CPUs

mgoudar opened the pull request:
Add scheduling model for the MIPS i6400 and i6500, an in-order MIPS64R6 processor.

i6400 and i6500 share same instruction latencies.

CPU has following pipelines
  - Two ALUs
  - Multiply and Divide unit (MDU)
  - Branch Unit (CTU)
  - Load/Store Unit (LSU)
  - Short Floating-point Unit and
  - Long Floating-point Unit
 
Latency information is available at: https://mips.com/wp-content/uploads/2025/03/MIPS_I6500-F_Data_Sheet_Rev1.10_3-17-2025.pdf

mgoudar committed:
[MIPS] Add Scheduling model for MIPS i6400 and i6500 CPUs

Add scheduling model for the MIPS i6400 and i6500, an in-order
MIPS64R6 processor.

i6400 and i6500 share same instruction latencies.

CPU has following pipelines
  - Two ALUs
  - Multiply and Divide unit (MDU)
  - Branch Unit (CTU)
  - Load/Store Unit (LSU)
  - Short Floating-point Unit and
  - Long Floating-point Unit

mshockwave commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,452 @@
+
+def MipsI6400Model : SchedMachineModel {
+  int IssueWidth = 2;         // 2x dispatched per cycle
+  int MicroOpBufferSize = 48; // min(48, 48, 64)
```
I6400 is an in-order processor right? Then I believe MicroOpBufferSize should be 0.

djtodoro replied:
It should be out-of-order.

mgoudar replied:
Hi, thanks for the catch. 16400 and i6500 are in-order processor. yes. sched model needs changes. I will make changes and update the pr.

mgoudar replied:
I have set MicroBufferSize and BufferSize to 0

mshockwave commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,452 @@
+    let ReleaseAtCycles = [5];
+  }
+  def I6400WriteLSULoad : SchedWriteRes<[I6400IssueLSU]> { let Latency = 3; }
+  def I6400WriteLSUPref : SchedWriteRes<[I6400IssueLSU]> { let Latency = 1; }
```
ditto

mgoudar replied:
done. removed 'let Latency = 1'

mshockwave commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,452 @@
+
+  // LSU pipelines
+  // =============
+  def : InstRW<[I6400WriteLSUStore], (instrs SB, SD, SH, SW, SDC1, SDC164, SWC1, SWC2_R6,
```
Is it possible avoid using `SchedWriteRes` + `InstRW` and using the normal `SchedWrite` + `WriteRes` instead?

`InstRW` really should be used for "ad-hoc fixing" nowadays (e.g. instruction A & B share the same scheduling class in their instruction definitions, but they should be in two different scheduling classes in your core)

mshockwave commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,452 @@
+
+  // FPU pipelines
+  // =============
+  def I6400FPU : ProcResource<3> { let BufferSize = 16; }
```
why the BufferSize is not zero for a resource in an in-order core?

djtodoro replied:
ditto

mgoudar replied:
initialized BufferSize for all resources to 0 as cpu is in-order core.

mshockwave commented on llvm/test/tools/llvm-mca/Mips/i6400.s:
```diff
@@ -0,0 +1,105 @@
+# NOTE: Assertions have been autogenerated by utils/update_mca_test_checks.py
+# RUN: llvm-mca -mtriple=mips64el -mcpu=i6400 -timeline -iterations=1 < %s | FileCheck %s
```
we don't usually need `-timeline` for scheduling model tests. Is there a specific reason you want to show timeline information here?

mshockwave replied:
> I wanted to test dispatch and execute timeline. do you suggest to remove --timeline from this test?

The scheduling model tests are supposed to check if the latency and occupancy (i.e. AcquireAtCycle / ReleaseAtCycle) show up correctly as expected. The basic MCA views already present those two metrics without `-timeline`. Testing with an additional `-timeline` wouldn't add much because it's nearly impossible for a scheduling model to "break" MCA's logics on how to calculate dispatch & execution cycles. That's why I would suggest removing `-timeline`

mgoudar replied:
I see. thank you for the information. I have removed timeline test now

mgoudar commented:
> Thanks for working on this! Can you please share a document that shows the instruction latencies?

Hi, please find the instruction latency at public document: https://mips.com/wp-content/uploads/2025/03/MIPS_I6500-F_Data_Sheet_Rev1.10_3-17-2025.pdf

mgoudar committed:
Set MicroOpBufferSize and BufferSize to 0

i6400 and i6500 are in-order cores. hence set BufferSize
and MicroBufferSize to 0

mgoudar commented on llvm/lib/Target/Mips/Mips32r6InstrInfo.td:
```diff
@@ -983,7 +983,10 @@ let AdditionalPredicates = [NotInMicroMips] in {
   def SWC2_R6 : SWC2_R6_ENC, SWC2_R6_DESC, ISA_MIPS32R6;
-  def SIGRIE : SIGRIE_ENC, SIGRIE_DESC, ISA_MIPS32R6;
+  
+  let hasNoSchedulingInfo = 1 in {
```
addressed the changes.

mgoudar commented on llvm/lib/Target/Mips/Mips32r6InstrInfo.td:
```diff
@@ -996,8 +999,10 @@ let AdditionalPredicates = [NotInMicroMips] in {
 let AdditionalPredicates = [NotInMicroMips] in {
-  def GINVI : R6MMR6Rel, GINVI_ENC, GINVI_DESC, ISA_MIPS32R6, ASE_GINV;
-  def GINVT : R6MMR6Rel, GINVT_ENC, GINVT_DESC, ISA_MIPS32R6, ASE_GINV;
+  let hasNoSchedulingInfo = 1 in {
```
addressed the changes.

mgoudar commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,452 @@
+  def I6400IssueLSU : ProcResource<1> { let Super = I6400AGEN; }
+  def I6400IssueALU1 : ProcResource<1> { let Super = I6400AGEN; }
+
+  def I6400WriteLSUStore : SchedWriteRes<[I6400IssueLSU]> { let Latency = 1; }
```
removed all instances where latency was set to 1

mshockwave commented on llvm/lib/Target/Mips/MipsScheduleI6400.td:
```diff
@@ -0,0 +1,447 @@
+
+  // AGEN Pipelines
+  // ==============
+  def I6400AGEN : ProcResource<1> { let BufferSize = 0; }
```
Setting `BufferSize = 0` on only some of the pipes is not enough I'm afraid. If this is an in-order core then every pipes should have `BufferSize = 0`. You can do it by
```
let BufferSize = 0 {
  def PipeA : ProcResource<1>;
  def PipeB : ProcResource<1>;
  ...
}
```

mshockwave commented on llvm/lib/Target/Mips/Mips32r6InstrInfo.td:
```diff
@@ -983,7 +983,10 @@ let AdditionalPredicates = [NotInMicroMips] in {
-  def SIGRIE : SIGRIE_ENC, SIGRIE_DESC, ISA_MIPS32R6;
+  
+  let hasNoSchedulingInfo = true in {
+    def SIGRIE : SIGRIE_ENC, SIGRIE_DESC, ISA_MIPS32R6;
```
if these instructions never have scheduling information. Why haven't other scheduling models complained about it (during TableGen generation time)?
Is it possible to set your new I6400 model with `CompleteModel = 0` instead of changing the instruction definition here?

mgoudar replied:
removed hasNoSchedulingInfo = true for unsupported instructions and set CompleteModel to 0 until we know resource and cycles for SIGRIE instruction.

Only generic model has set latency to 1 for this instruction. So no issues there.

mgoudar committed:
Address review comments

- remove hasNoSchedulingInfo in Mips32r6InstrInfo.td
- define BufferSize to 0 for all resources
- remove timeline test in llvm-mca testcase

mshockwave commented on llvm/test/tools/llvm-mca/Mips/i6400.s:
```diff
@@ -0,0 +1,93 @@
+	mul	$3, $2, $3
+	sw	$3, 8($sp)
+	div	$2, $1
+	sw	$2, 0($sp)
```
Could you add more tests on simple integer and floating point arithmetics instructions, as well as some vector instructions (they're called MSA?).

mgoudar replied:
yes. I will and. thank you for the review

mgoudar committed:
Cover more instructions in llvm-mca test

Extend i64..s to cover more instructions including MSA

mgoudar committed:
Update llvm-mca test and rerun update_mca_test_checks.py

removed --timeline option from the RUN command

mgoudar commented on llvm/test/tools/llvm-mca/Mips/i6400.s:
```diff
@@ -0,0 +1,133 @@
+# NOTE: Assertions have been autogenerated by utils/update_mca_test_checks.py
+# RUN: llvm-mca -mtriple=mips64el -mcpu=i6400 --mattr=msa -timeline -iterations=1 < %s | FileCheck %s
```
thank you for pointing me the process. I have now removed --timeline option and rerun with update_mca_test_checks.py

mgoudar commented on llvm/test/tools/llvm-mca/Mips/i6400.s:
```diff
@@ -0,0 +1,133 @@
+# NOTE: Assertions have been autogenerated by utils/update_mca_test_checks.py
+# RUN: llvm-mca -mtriple=mips64el -mcpu=i6400 --mattr=msa -timeline -iterations=1 < %s | FileCheck %s
```
i6400 and i6500 comes with MSA implemented and enabled by default, user can disable MSA with reset bit from the software. I think I need to enable FeatureMSA as part of i6400/i6500 cpu definition. I shall create  a separate PR for that and share here.

mshockwave reviewed (approved):
Personally I'm not really a fan of using InstRW as I pointed out in one of my comments. But I guess it's up to MIPS maintainers to make the call on that. Other than that, LGTM.
@djtodoro do you have any other comments?

djtodoro reviewed (changes requested):
Lets try to avoid `InstRW`.

mgoudar commented:
> Lets try to avoid `InstRW`.

Hi, thank you for the review and the feedback. We have started porting work. We will port all existing scheduling model for MIPS cpus to new scheduling API soon.

djtodoro commented:
OK, it can be addressed as a separate PR, but please handle all MIPS CPUs that use `InstRW`. Thanks!

mgoudar commented:
> OK, it can be addressed as a separate PR, but please handle all MIPS CPUs that use `InstRW`. Thanks!

yes sure. thank you.

mshockwave commented:
> LLVM Buildbot has detected a new failure on builder `llvm-nvptx-nvidia-win` running on `as-builder-8` while building `llvm` at step 7 "test-build-unified-tree-check-llvm".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/54/builds/8671
> 
> Here is the relevant piece of the build log for the reference

You forgot to add `lit.local.cfg` like this: https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/llvm-mca/RISCV/lit.local.cfg
Which limits the tests to LLVM that has the MIPS backend (or you can add `REQUIRES: mips-registered-target` to every single tests under the Mips folder, but usually people prefer lit.local.cfg)

mgoudar commented:
> > LLVM Buildbot has detected a new failure on builder `llvm-nvptx-nvidia-win` running on `as-builder-8` while building `llvm` at step 7 "test-build-unified-tree-check-llvm".
> > Full details are available at: https://lab.llvm.org/buildbot/#/builders/54/builds/8671
> > Here is the relevant piece of the build log for the reference
> 
> You forgot to add `lit.local.cfg` like this: https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/llvm-mca/RISCV/lit.local.cfg Which limits the tests to LLVM that has the MIPS backend (or you can add `REQUIRES: mips-registered-target` to every single tests under the Mips folder, but usually people prefer lit.local.cfg)

Thank you. I will add lit.local.cfg and create new pr. thanks for your feedback.
